everestrs/everestrs/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+
+        /// Registers the callback of the `GenericModule` to be called and calls
+        /// `Everest::Module::signal_ready`.
+        unsafe fn signal_ready(
+            self: &Module,
+            // See comment in `CommandMeta::obj`.
+            obj: *const c_char,
+            on_ready: unsafe fn(obj: *const c_char) -> (),
+        );
+
+        /// Informs the runtime that we implement the command described in `meta` and registers the
+        /// `handle_command` method from the `GenericModule` as the handler.
+        unsafe fn provide_command(
+            self: &Module,
+            meta: CommandMeta,
+            on_ready: unsafe fn(&CommandMeta, JsonBlob) -> JsonBlob,
+        );
+    }
+}
+
+impl ffi::JsonBlob {
+    fn as_bytes(&self) -> &[u8] {
+        &self.data
+    }
+
+    fn deserialize<T: DeserializeOwned>(self) -> T {
+        // TODO(hrapp): Error handling
+        serde_json::from_slice(self.as_bytes()).unwrap()
+    }
+
+    fn from_vec(data: Vec<u8>) -> Self {
+        Self { data }
+    }
+}
+
+#[derive(FromArgs, Debug)]
+/// An everest Node.
+struct Args {
+    /// prefix of installation.
+    #[argh(option)]
+    #[allow(unused)]
+    pub prefix: PathBuf,
+
+    /// configuration yml that we are running.
+    #[argh(option)]
+    #[allow(unused)]
+    pub conf: PathBuf,
+
+    /// module name for us.
+    #[argh(option)]
+    pub module: String,
+}
+
+/// Implements the handling of commands & variables, but has no specific information about the
+/// details of the current module, i.e. it deals with JSON blobs and strings as command names. Code
+/// generation is used to build the concrete, strongly typed abstractions that are then used by
+/// final implementors.
+pub trait GenericModule: Sync {
+    /// Handler for the command `name` on `implementation_id` with the given `parameters`. The return value
+    /// will be returned as the result of the call.
+    fn handle_command(
+        &self,
+        implementation_id: &str,
+        name: &str,
+        parameters: HashMap<String, serde_json::Value>,
+    ) -> Result<serde_json::Value>;
+
+    fn on_ready(&self) {}
+}
+
+pub struct Runtime<T: GenericModule> {
+    // There are two subleties here:
+    // 1. We are handing out pointers to `module_impl` to `cpp_module` for callbacks. The pointers
+    //    must must stay valid for as long as `cpp_module` is alive. Hence `module_impl` must never
+    //    move in memory. Rust can model this through the Pin concept which upholds this guarantee.
+    //    We use a Box to put the object on the heap.
+    // 2. For the same reason, `module_impl` should outlive `cpp_module`, hence should be dropped
+    //    after it. Rust drops fields in declaration order, hence `cpp_module` should come before
+    //    `module_impl` in this struct.
+    cpp_module: cxx::UniquePtr<ffi::Module>,
+    module_impl: Pin<Box<T>>,
+}
+
+impl<T: GenericModule> Runtime<T> {
```

Comment:
no, it was not necessary, since we anyways only capture a pointer to the heap to the module, it is anyways always used with dynamic dispatch, so we can drop the templating. @dorezyuk pointed that out too. done now.